samplejava/src/main/java/com/example/chattutorial/ChannelActivity3.java

```diff
@@ -107,29 +120,43 @@ public void handleOnBackPressed() {
         });
 
         // Custom typing info header bar
-        TextView typingHeaderView = findViewById(R.id.typingHeaderView);
         String nobodyTyping = "nobody is typing";
-        typingHeaderView.setText(nobodyTyping);
-
-        // Obtain a ChannelController
-        ChatDomain.instance()
-                .getChannelController(cid)
-                .enqueue((result) -> {
-                    ChannelController channelController = result.data();
-
-                    // Observe typing users
-                    channelController.getTyping().observe(this, typingState -> {
-                        if (typingState.getUsers().isEmpty()) {
-                            typingHeaderView.setText(nobodyTyping);
-                        } else {
-                            List<String> userNames = new LinkedList<>();
-                            for (User user : typingState.getUsers()) {
-                                userNames.add(user.getName());
-                            }
-                            String typing = "typing: " + TextUtils.join(", ", userNames);
-                            typingHeaderView.setText(typing);
-                        }
-                    });
-                });
+        binding.typingHeaderView.setText(nobodyTyping);
+
+        // Observe typing events and update typing header depending on its state.
+        Flow<ChannelState> channelStateFlow = ChatClientExtensions.watchChannelAsState(ChatClient.instance(), cid, 30);
```

Comment:
This is very clunky, but we currently don't expose any of our state classes (`ChannelState`, `GlobalState`, ...) as `LiveData`.

It was either this or suggesting that Java users use Coroutines, which I feel is a bigger unknown in the Java world + the syntax would still be clunky since `Flow` operators are written as extensions.